git/odb/commit.go

```diff
@@ -0,0 +1,216 @@
+	// signature.
+	Name string
+	// Email is the email address of the individual holding this signature.
+	Email string
+	// When is the instant in time when the signature was created.
+	When time.Time
+}
+
+const (
+	parseTimeZoneOnly  = "2006 -0700"
+	formatTimeZoneOnly = "-0700"
+)
+
+// String implements the fmt.Stringer interface and formats a Signature as
+// expected in the Git commit internal object format. For instance:
+//
+//  Taylor Blau <ttaylorr@github.com> 1494258422 -0600
+func (s *Signature) String() string {
+	at := s.When.Unix()
+	zone := s.When.Format(formatTimeZoneOnly)
+
+	return fmt.Sprintf("%s <%s> %d %s", s.Name, s.Email, at, zone)
+}
+
+var (
+	signatureNameRe  = regexp.MustCompile("^[^<]+")
+	signatureEmailRe = regexp.MustCompile("<(.*)>")
+	signatureTimeRe  = regexp.MustCompile("[-]?\\d+")
+)
+
+// ParseSignature parses a given string into a signature instance, returning any
+// error that it encounters along the way.
+//
+// ParseSignature expects the signature encoded in the given string to be
+// formatted correctly, and reproduce-able by the Signature.String() function
+// above.
+func ParseSignature(str string) (*Signature, error) {
+	name := signatureNameRe.FindString(str)
+	emailParts := signatureEmailRe.FindStringSubmatch(str)
+	timeParts := signatureTimeRe.FindAllStringSubmatch(str, 2)
+
+	if len(emailParts) < 2 {
+		return nil, errors.Errorf("git/odb: expected email in signature: %q", str)
+	}
+	email := emailParts[1]
+
+	if len(timeParts) < 1 {
+		return nil, errors.Errorf("git/odb: expected time in signature: %q", str)
+	}
+
+	epoch, err := strconv.ParseInt(timeParts[0][0], 10, 64)
+	if err != nil {
+		return nil, err
+	}
+
+	t := time.Unix(epoch, 0).In(time.UTC)
+	if len(timeParts) > 1 {
+		loc, err := parseTimeZone(timeParts[1][0])
+		if err != nil {
+			return nil, err
+		}
+
+		t = t.In(loc)
+	}
+
+	return &Signature{
+		Name:  strings.TrimSpace(name),
+		Email: email,
+		When:  t,
+	}, nil
+}
+
+// parseTimeZone returns the *time.Location corresponding to a Git-formatted
+// string offset. For instance, the string "-0700" would format into a
+// *time.Location that is 7 hours east of UTC.
+func parseTimeZone(zone string) (*time.Location, error) {
+	loc, err := time.Parse(parseTimeZoneOnly, fmt.Sprintf("1970 %s", zone))
+	if err != nil {
+		return nil, err
+	}
+	return loc.Location(), nil
+}
+
+// Commit encapsulates a Git commit entry.
+type Commit struct {
+	// Author is the Author this commit, or the original writer of the
+	// contents.
+	Author *Signature
+	// Committer is the individual or entity that added this commit to the
+	// history.
+	Committer *Signature
+	// ParentIds are the IDs of all parents for which this commit is a
+	// linear child.
+	ParentIds [][]byte
+	// TreeId is the root Tree associated with this commit.
+	TreeId []byte
+	// Message is the commit message, including any signing information
+	// associated with this commit.
+	Message string
+}
+
```

Comment:
You may encounter other headers besides these. At the very least, Git may generate `encoding`, `gpgsig`, and `mergetag` headers. It's not encouraged, but you could also see other random headers added by other implementations.

You'd want to preserve these in general (though for something like `gpgsig`, you'd eventually want to support re-signing). And you'd probably want to do so as byte-for-byte identically as you can, just replacing the parent/tree sha1 pointers as needed.

It looks like your approach is to do a Decode/Modify/Encode pair for each object. While that's pleasant to work with during the modify step, I think you may want to consider instead doing a line-by-line filter, where you rewrite the lines whose sha1s are modified, and otherwise try to keep everything byte-for-byte identical.